### DISCUSSION TITLE
Is it good to NOT run adopt_or_reset_orphaned_tasks on startup ?

### DISCUSSION THREAD

**Discussion by NickYadance** [apache/airflow#27983] [Ideas]
The first `adopt_or_reset_orphaned_tasks` run is slow(~10sec) when restarting scheduler due to lots of ohphaned tasks generated, especially with KubernetesExecutor, which is making serial HTTP calls to adopt those tasks. And as `adopt_or_reset_orphaned_tasks` can potentially lock all the orphaned task instances, multiple scheduler startup will be stuck for quite some time with database not supporting skip_lock. 

Not sure what side effects can have by keeping these ohphaned tasks while the scheduler is running, is it good to avoid it on startup ?

**Comment by potiuk** [comment]
You are not supposed to run multiple schedulers without SKIP_LOCK. You will have all kinds of weird errors if you do:

https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html

> MariaDB did not implement the SKIP LOCKED or NOWAIT SQL clauses until version [10.6.0](https://jira.mariadb.org/browse/MDEV-25433). Without these features, running multiple schedulers is not supported and deadlock errors have been reported. MariaDB 10.6.0 and following may work appropriately with multiple schedulers, but this has not been tested.

I suggest you upgrade to a DB that is supported and your problem will be gone.

**Reply by NickYadance** [reply]
thx for the reply. i am running multiple schedulers with mysql 8.1. 

The main reason behind slow running adopt_or_reset_orphaned_tasks is lots of (serialized) http calls behind KubernetesExecutor adopting tasks. For KubernetesExecutor, it runs adoption by searching all the pods by scheduler_id of the last Scheduler instance, which can be a lot for large clusters. 

I'll check if this is the MUST thing for scheduler, perhaps the scheduler will generate duplicated task instances without this.
https://github.com/apache/airflow/blob/1aa3da543a3f9229527a5de2807053e15b2bfea7/airflow/jobs/scheduler_job.py#L839-L840

**Reply by potiuk** [reply]
Ah. You mentioned "multiple schedulers" and "no skip lock" as a problem - I thought **this** was the problem.

The reason is simple. Scheduler, when running keeps information about the tasks running via K8S executor in memory. The "adppt orphaned tasks" will simply restore the  in-memory state for the tasks to monitor their state while they are running. Scheduler will need the state to manage failure/recovery in case the tasks fail to update their state in database.

So yes - this is neeed and yes eventually it might lead to duplications.

Maybe some improvements can be done to speed it up - maybe we could run the adoption in multiple parallell threads or even in parallell to regular scheduler loop? I think in general the problem is that we might exceed some of the limitations (pools, parallelism) for number of running tasks but maybe it's worth it. 

That is I think similar problem to one described (and possibly solved) in https://github.com/apache/airflow/pull/26639  

Seems that when there are many K8S-executor controlled tasks, serialization of the K8S API becomes a bottleneck for some housekeeping tasks as it is linear and sequential - and in many cases where K8S cluster is remote, the round-trip and SSL times might easily add up to cause even long delays.

@dstandish @jedcunningham @ephraimbuddy - what do you think? I think both https://github.com/apache/airflow/pull/26639 and this one might be viable optimizations. 

Do you see any drawbacks of either of those ?

**Reply by NickYadance** [reply]
To avoid k8s apiserver backpressure, some parallelism configurations will be needed too, if this goes in a parallel way, as in #26639.

**Reply by potiuk** [reply]
Yep.

**Reply by hterik** [reply]
In https://github.com/apache/airflow/issues/28071 there is suggestion of a solution to exclude already completed tasks from adoption, that would save a lot of time in that process.

**Reply by NickYadance** [reply]
In my deployment, i have configured Airflow to delete pods of completed tasks so it won't work for me. 
@hterik

**Reply by NickYadance** [reply]
As for #26639, i'm trying to utilize the underground `api_client` of `kube_client`, which already offers nice support of async request, instead of constructing threadpool manually.
